### PR TITLE
Plat 5040 - New FSM and register/unregister from live proxy

### PR DESCRIPTION
**Pull request by gadyaari**

### CODE REVIEW

File: common/config/config.json.template [hunk 1]
```diff
@@ -6,12 +6,14 @@
   "oldContentFolderPath" : "/dvrOldContentRootPath",
   "requestTimeout" : 20000,
   "pollingInterval" : 2000,
-  "backendPollingInterval" : 3000,
+  "backendPollingInterval" : 5000,
```
Comment (kalturaguy) [inline common/config/config.json.template]: add new one-> wowzaPollingInterval

File: lib/Adapters/WowzaAdapter.js [hunk 2]
```diff
@@ -5,78 +5,92 @@
...
 
-
-var applicationName=config.get('mediaServer').applicationName;
+var applicationName = config.get('mediaServer').applicationName;
 var hostname = config.get('mediaServer').hostname;
 var port = config.get('mediaServer').port;
-var wowzaUrl="http://localhost:8087/v2/servers/_defaultServer_/vhosts/_defaultVHost_/applications/"+applicationName+"/instances";
-//var wowzaUrl2="http://localhost:8087/v2/servers/_defaultServer_/vhosts/_defaultVHost_/applications/"+applicationName+"/instances/_definst_/incomingstreams/"+entryId+"/monitoring/current”
-var backendClient=require('../BackendClientFactory.js').getBackendClient();
+var wowzaUrl = "http://" + hostname + ":8087/v2/servers/_defaultServer_/vhosts/_defaultVHost_/applications/" + applicationName + "/instances";
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: please add user+password support both here + configuration

File: lib/Adapters/WowzaAdapter.js [hunk 3]
```diff
@@ -5,78 +5,92 @@
...
+           return Q.all(promises)
+               .then(function(res) {
+                    return Q.resolve(res);
+           });
 
 
         })
         .catch(function(err) {
-            console.warn(err,"",err.stack);
+            console.warn("Received the following warning: " + err, "", err.stack);
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: change console.warn to logger

File: lib/Adapters/WowzaAdapter.js [hunk 4]
```diff
@@ -5,78 +5,92 @@
...
+
+WowzaAdapter.prototype.getLiveEntries = function() {
+    var that = this;
+    return networkClient.read({
+            url : wowzaUrl,
+            json : true
         })
         .then(function(res) {
-            if (res.instanceList.length==0) {
+            if (res.instanceList.length === 0) {
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: please add logger!!

File: lib/Adapters/WowzaAdapter.js [hunk 5]
```diff
@@ -5,78 +5,92 @@
...
+        return tmp;
+    });
+    _.each(unactiveEntries, function(e) {
 
-WowzaAdapter.prototype.getLiveEntries=function() {
-    var self=this;
-    return NetworkClient.read({
-            url:wowzaUrl,
-            json:true
+        delete that.entries[e.entryId];
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: add logs

File: lib/Adapters/WowzaAdapter.js [hunk 6]
```diff
@@ -5,78 +5,92 @@
...
-                        newEntry.sourceIp = stream.sourceIp;
-                        newEntry.getStreamInfo = getStreamInfo;
-                        return Q.resolve(newEntry);
+                    .then(function(entryArr) {
+                        if (entryArr.length) {
+                            var newEntry = entryArr[0];
+                            that.entries[entryId] = newEntry;
+                            newEntry.sourceIp = stream.sourceIp;
+                            newEntry.getStreamInfo = getStreamInfo;
+                            return Q.resolve(newEntry);
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: add to log

File: lib/Adapters/WowzaAdapter.js [hunk 7]
```diff
@@ -5,78 +5,92 @@
...
+                        }
+                        return Q.reject("No entries received");
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: add logs

File: lib/Adapters/WowzaAdapter.js [hunk 8]
```diff
@@ -5,78 +5,92 @@
...
                 });
             });
 
-           return Q.all(promises).then(function(res) {
-               return Q.resolve(res);
-           })
+           return Q.all(promises)
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: check what happens if 1 rjected and reset OK

File: lib/BackendClient.js [hunk 9]
```diff
@@ -13,13 +13,13 @@ var logger = require('./logger/logger')(module);
...
 
-    var createClientConfig = function () {
+    function createClientConfig() {
         var conf = new kalturaClientBase.KalturaConfiguration(partnerId);
         conf.serviceUrl = config.get('backendClient').serviceUrl;
         return conf;
     };
 
-    var convertToMs = function (minutes) {
+    var convertToMs = function(minutes) {
```
Comment (kalturaguy) [inline lib/BackendClient.js]: rename function

File: lib/BackendClient.js [hunk 10]
```diff
@@ -27,13 +27,13 @@ var BackendClient = (function(){
     var adminSecret = config.get('backendClient').adminSecret;
```
Comment (kalturaguy) [inline lib/BackendClient.js]: for every API call please add response headers + actual response to log

File: lib/BackendClient.js [hunk 11]
```diff
@@ -116,54 +117,124 @@ var BackendClient = (function(){
...
-    };
+    }
 
-    BackendClient.getLiveEntriesForMediaServer = function () {
-       return createSession().
-            then(function () {
+    function registerEntryInDatabase(entryId, serverIndex, state, event) {
+        var deferred = Q.defer();
+        var startTime = new Date();
+        logger.debug("Request for register media server, started at " + startTime);
```
Comment (kalturaguy) [inline lib/BackendClient.js]: no need to double print the time

File: lib/BackendClient.js [hunk 12]
```diff
@@ -116,54 +117,124 @@ var BackendClient = (function(){
...
+        logger.info("Calling unregisterMediaServer, and delete entry %s from database", entryId);
+        return createSession()
+            .then(function() {
+                return unregisterEntryInDatabase(entryId);
+            });
+
+
+    };
+    //This method is use for debugging ---NOT WORKING
+    BackendClient.IsLive = function(entryId) {
```
Comment (kalturaguy) [inline lib/BackendClient.js]: please remove if not used

File: lib/entry/FlavorDownloader.js [hunk 13]
```diff
@@ -276,29 +263,43 @@ var FlavorDownloader = (function () {
...
+                            var downloadChunksPromise = downloadChunks.call(that, baseUrl, Object.keys(that.chunksToDownload));
                             return Q.allSettled([downloadChunksPromise, m3u8generatorPromise]);
-                        }).then(function (results) {
+                        })
+                        .then(function(results) {
+                            // TODO: What if chunklist hasn't changed since last iteration? -> Gad
+                            var tmp = that.m3u8generator.getCurrentManifest();
+                            if (tmp.items.PlaylistItem.length > 0) {
+                                that.logger.info("Finished downloading new chunks, raising 'newTsFiles' event");
+                                that.emit('newTsFiles', tmp.items.PlaylistItem.length);
```
Comment (kalturaguy) [inline lib/entry/FlavorDownloader.js]: please change the event to have in parameters the list of new chunks, deleted chunks and list of current chunks

File: lib/entry/FlavorDownloader.js [hunk 14]
```diff
@@ -342,45 +344,27 @@ var FlavorDownloader = (function () {
...
-        var teardown = Q();
-        if (that.runStatus === 'running')
-        {
-            var teardownInterval = config.get("flavorDownloaderTeardownInterval");
-            that.logger.info('stop requested - allowing ' + teardownInterval + 'ms for orderly tear-down');
-            teardown = Q.delay(teardownInterval).then(function(){
-                that.logger.info('stopping flavor downloader after allowing orderly tear-down');
-            });
-        }
+        return d.promise;
```
Comment (kalturaguy) [inline lib/entry/FlavorDownloader.js]: FlavorDownloader.prototype.stop = function () {
        var that = this;
        that.logger.info("Stopping flavor downloader");
        that.shouldRun = false;
        return downloaderEnd.promise.then(function() {
            that.logger.info("FlavorDownloader Stopped");
            delete FlavorDownloader.activeInstances[that.id];
            that.runStatus = 'stopped';
            that.emit('stopped', that.flavor);
            
        });


    };

File: lib/Worker.js [hunk 15]
```diff
@@ -72,46 +72,47 @@ module.exports = (function(){
...
+                        });
                 }
             });
     };
 
     var processAlreadyLiveEntries = function(currentlyLiveEntries) {
         var that = this;
-        var alreadyLiveEntries = _.filter(currentlyLiveEntries, function (e) {
+        var alreadyLiveEntries = _.filter(currentlyLiveEntries, function(e) {
             var entryIds = Object.keys(that.handledEntries);
```
Comment (kalturaguy) [inline lib/Worker.js]: use _.has instead

File: lib/Worker.js [hunk 16]
```diff
@@ -126,31 +127,29 @@ module.exports = (function(){
         _.forEach(unhandledLiveEntries, function (e) {
 
             // Mark that this entry is already being handled (by the "pending" indicator) so that another
-            // downloader won't be created for it until the previous one's creation has completed
+            // entry won't be created for it until the previous one's creation has completed
```
Comment (kalturaguy) [inline lib/Worker.js]: use _.has instead above

File: lib/Worker.js [hunk 17]
```diff
@@ -126,31 +127,29 @@ module.exports = (function(){
         _.forEach(unhandledLiveEntries, function (e) {
```
Comment (kalturaguy) [inline lib/Worker.js]: rename to processNewLiveEntries

File: lib/Worker.js [hunk 18]
```diff
@@ -126,31 +127,29 @@ module.exports = (function(){
...
 
             // Mark that this entry is already being handled (by the "pending" indicator) so that another
-            // downloader won't be created for it until the previous one's creation has completed
+            // entry won't be created for it until the previous one's creation has completed
             if (!that.handledEntries[e.entryId] && (!that.phantomEntries[e.entryId] || that.phantomEntries[e.entryId] < maxEntryInitializationTryCount))
             {
```
Comment (kalturaguy) [inline lib/Worker.js]: fix { (move to line above)

File: lib/Worker.js [hunk 19]
```diff
@@ -173,53 +172,49 @@ module.exports = (function(){
...
-                .done();
+                // TODO: Stop log writing from being called so many times! -> Gad
+                logger.info("Entry %s is no longer live - calling liveEntry.stop", noLongerLiveEntry);
+                return liveEntry.stop()
+                    .then(function(result) {
+                        if (result) {
+                            logger.info("Entry %s is no longer live - removing it from the list of handled entries", noLongerLiveEntry);
+                            delete that.handledEntries[noLongerLiveEntry];
+                            logger.debug("Adding entry %s to endlistAppender", noLongerLiveEntry);
+                            return that.endListAppender.addEntry(noLongerLiveEntry) // Indicate entry is down and should (later) have end-list appended
```
Comment (kalturaguy) [inline lib/Worker.js]: no sense into keeping it here

File: lib/entry/LiveEntry.js [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+function LiveEntry(entryObject, isNewSession) {
+    events.EventEmitter.call(this);
+    this.entryObject = entryObject;
+    this.entryId = entryObject.entryId;
+    this.masterManifestGenerator = new masterManifestCtor(entryObject, isNewSession);
+    this.manifestTimeWindow = entryObject.manifestTimeWindow;
+    this.manifestMaxChunkCount = entryObject.maxChunkCount;
+    this.flavorsDownloaders = [];
+    this.FSM = FSM(entryObject);
+    this.playbackRaised = false;
```
Comment (kalturaguy) [inline lib/entry/LiveEntry.js]: no need for it

File: lib/entry/StateManager.js [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+        // TODO: Check difference between stop from BROADCASTING to stop from PLAYING -> Gad
+        lastRegisteredInAPI = null;
+        logger.info("Event '%s' raised: Changing state from '%s' to '%s'", event.toUpperCase(), from, to);
+        return backendClient.unregisterEntryInDatabase(entryObject.entryId);
+    }
+
+    function onplay(event, from, to) {
+        // TODO: Erase duplicate code!!! -> Gad
+        switch (event) {
+            case "play":
```
Comment (kalturaguy) [inline lib/entry/StateManager.js]: remove duplication

File: lib/entry/StateManager.js [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+                logger.info("Event '%s' raised: Changing state from '%s' to '%s'", event.toUpperCase(), from, to);
+                // Refresh entry's timestamp
+                return sessionManager.refreshSessionTimestamp(entryObject.entryId)
+                    .then(function() {
+                        lastRegisteredInAPI = new Date().getTime();
+                        return backendClient.registerEntryInDatabase(entryObject, KalturaLiveStatus.PLAYABLE, event)
+                            .catch(function(err) {
+                                logger.error("Failed to register media server: ", err);
```
Comment (kalturaguy) [inline lib/entry/StateManager.js]: add %s

File: lib/entry/StateManager.js [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+            onplay : onplay,
+            onupdate : onplay
+        }
+    });
+
+    function onstart(event, from, to) {
+        // TODO: Insert ability to move from 'init' state to 'playing' immediately? -> Gad
+        logger.info("Event '%s' raised: Changing state from '%s' to '%s'", event.toUpperCase(), from, to);
+        return backendClient.registerEntryInDatabase(entryObject, KalturaLiveStatus.BROADCASTING, event);
+    }
```
Comment (kalturaguy) [inline lib/entry/StateManager.js]: add log if failed

File: lib/entry/StateManager.js [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+
+    function onstop(event, from, to) {
+        // TODO: Check difference between stop from BROADCASTING to stop from PLAYING -> Gad
+        lastRegisteredInAPI = null;
+        logger.info("Event '%s' raised: Changing state from '%s' to '%s'", event.toUpperCase(), from, to);
+        return backendClient.unregisterEntryInDatabase(entryObject.entryId);
+    }
```
Comment (kalturaguy) [inline lib/entry/StateManager.js]: add log if failed

File: common/config/config.json.template (hunk 1 above)
Comment (gadyaari) [inline common/config/config.json.template]: done

File: lib/Adapters/WowzaAdapter.js (hunk 2 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 5 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 4 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 6 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 8 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 3 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/BackendClient.js (hunk 9 above)
Comment (gadyaari) [inline lib/BackendClient.js]: done

File: lib/BackendClient.js (hunk 11 above)
Comment (gadyaari) [inline lib/BackendClient.js]: done

File: lib/BackendClient.js (hunk 12 above)
Comment (gadyaari) [inline lib/BackendClient.js]: done

File: lib/Worker.js (hunk 15 above)
Comment (gadyaari) [inline lib/Worker.js]: done

File: lib/Worker.js (hunk 17 above)
Comment (gadyaari) [inline lib/Worker.js]: done

File: lib/Worker.js (hunk 18 above)
Comment (gadyaari) [inline lib/Worker.js]: done

File: lib/Worker.js (hunk 19 above)
Comment (gadyaari) [inline lib/Worker.js]: Was decided to leave it in Worker for now

File: lib/entry/FlavorDownloader.js (hunk 14 above)
Comment (gadyaari) [inline lib/entry/FlavorDownloader.js]: done

File: lib/entry/StateManager.js (hunk 23 above)
Comment (gadyaari) [inline lib/entry/StateManager.js]: done

File: lib/entry/StateManager.js (hunk 24 above)
Comment (gadyaari) [inline lib/entry/StateManager.js]: done

File: lib/entry/StateManager.js (hunk 22 above)
Comment (gadyaari) [inline lib/entry/StateManager.js]: done

File: lib/entry/StateManager.js (hunk 21 above)
Comment (gadyaari) [inline lib/entry/StateManager.js]: done

File: lib/entry/LiveEntry.js (hunk 20 above)
Comment (gadyaari) [inline lib/entry/LiveEntry.js]: debatable

File: lib/entry/FlavorDownloader.js (hunk 13 above)
Comment (gadyaari) [inline lib/entry/FlavorDownloader.js]: On hold for now. Waiting for integration with Igor and his changes

File: lib/Worker.js (hunk 16 above)
Comment (gadyaari) [inline lib/Worker.js]: done

File: lib/BackendClient.js (hunk 10 above)
Comment (gadyaari) [inline lib/BackendClient.js]: done

File: lib/Adapters/WowzaAdapter.js [hunk 25]
```diff
@@ -6,20 +6,23 @@ var Q = require('q');
...
+
 
 var applicationName = config.get('mediaServer').applicationName;
 var hostname = config.get('mediaServer').hostname;
 var port = config.get('mediaServer').port;
-var wowzaUrl = "http://" + hostname + ":8087/v2/servers/_defaultServer_/vhosts/_defaultVHost_/applications/" + applicationName + "/instances";
+var user = config.get('mediaServer').user;
+var password = config.get('mediaServer').password;
+var prefix = ((user !== "") && (password !== "")) ? user + ":" + password + "@" + hostname : hostname;
+var wowzaUrl = "http://" + prefix + ":8087/v2/servers/_defaultServer_/vhosts/_defaultVHost_/applications/" + applicationName + "/instances";
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: "http://" + authentication + host+ ":8087/

File: lib/Adapters/WowzaAdapter.js [hunk 26]
```diff
@@ -6,20 +6,23 @@ var Q = require('q');
 var _ = require('underscore');
 var config = require('../../common/Configuration');
 var networkClient = require('../NetworkClient.js');
-var url = require('url');
+var logger = require('../logger/logger')(module);
 var baseAdapter = require('./BaseAdapter.js').BaseAdapter;
+var wowzaStreamInfo = require('./WowzaStreamInfo.js');
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: W and not w

File: lib/Adapters/WowzaAdapter.js [hunk 27]
```diff
@@ -49,7 +53,8 @@ WowzaAdapter.prototype.getLiveEntries = function() {
             json : true
         })
         .then(function(res) {
-            if (res.instanceList.length === 0) {
+            logger.debug("Received playing entries list from Wowza")
+            if (res.instanceList[0].incomingStreams.length === 0) {
                 return [];
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: add log here

File: lib/Adapters/WowzaAdapter.js [hunk 28]
```diff
@@ -62,7 +67,8 @@ WowzaAdapter.prototype.getLiveEntries = function() {
                 return entryId
             });
             // Make sure module's entries list doesn't contain entries that are no longer streaming
-            deleteUnactiveStreams.call(that, uniqueList);
+            logger.debug("Check for streams that no longer play and update wowzaAdapter's entries list");
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: duplication of log line

File: lib/Adapters/WowzaAdapter.js [hunk 29]
```diff
@@ -27,17 +30,18 @@ function WowzaAdapter() {
...
-        var tmp = _.findWhere(currentlyActiveStreams, {'entryId' : e1.entryId});
-        return tmp;
+    logger.debug("Retrieve list of entries that are no longer active for deletion");
+    var unactiveEntries = _.reject(that.entries, function(e1) {
+        return _.findWhere(currentlyActiveStreams, {'entryId' : e1.entryId});
     });
+    logger.debug("Deleting %d entries from list", unactiveEntries.length);
     _.each(unactiveEntries, function(e) {
-
+        logger.debug("Entry %s deleted", e.entryId);
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: deleting entry %s

File: lib/Adapters/WowzaAdapter.js [hunk 30]
```diff
@@ -27,17 +30,18 @@ function WowzaAdapter() {
...
+function deleteInactiveStreams(currentlyActiveStreams) {
     var that = this;
-    var unactiveEntries = _.reject(that.entries, function(e1) {;
-        var tmp = _.findWhere(currentlyActiveStreams, {'entryId' : e1.entryId});
-        return tmp;
+    logger.debug("Retrieve list of entries that are no longer active for deletion");
+    var unactiveEntries = _.reject(that.entries, function(e1) {
+        return _.findWhere(currentlyActiveStreams, {'entryId' : e1.entryId});
     });
+    logger.debug("Deleting %d entries from list", unactiveEntries.length);
```
Comment (kalturaguy) [inline lib/Adapters/WowzaAdapter.js]: remove

File: lib/BackendClient.js [hunk 31]
```diff
@@ -33,25 +33,33 @@ var BackendClient = (function(){
     var sessionDuration = convertToMs(config.get('backendClient').ksSessionRefreshIntervalMinutes);
     var BackendClient = {};
 
+    function printAPIResponse(result, headers) {
+        logger.info("API call headers: %j", headers);
```
Comment (kalturaguy) [inline lib/BackendClient.js]: print only important headers
+ print result only if fails

File: lib/Worker.js [hunk 32]
```diff
@@ -101,74 +101,70 @@ module.exports = (function(){
...
-                logger.info("Entry " + e.entryId + " started playing - creating an entry object for it");
-                createLiveEntry.call(that, e).then(
-                    function() {
+                logger.info("Entry %s started playing - creating an entry object for it", e.entryId);
+                createLiveEntry.call(that, e)
+                    .then(function() {
                         // If the entry object creation is successful remove it from the list of phantom entries
                         if (that.phantomEntries[e.entryId]) {
-                            logger.info('Successfully created liveEntry object for entry ' + e.entryId + ' removing it from the phantom list');
+                            logger.info("Successfully created liveEntry ob  ject for entry %s removing it from the phantom list", e.entryId);
```
Comment (kalturaguy) [inline lib/Worker.js]: i ob           ject

File: lib/Adapters/WowzaAdapter.js (hunk 26 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 25 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 30 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 29 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 27 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/Adapters/WowzaAdapter.js (hunk 28 above)
Comment (gadyaari) [inline lib/Adapters/WowzaAdapter.js]: done

File: lib/BackendClient.js (hunk 31 above)
Comment (gadyaari) [inline lib/BackendClient.js]: done

File: lib/Worker.js (hunk 32 above)
Comment (gadyaari) [inline lib/Worker.js]: done

**Pull request closed without merging**